madmom/evaluation/chords.py

```diff
@@ -0,0 +1,897 @@
+
+CHORD_ANN_DTYPE = [('start', np.float),
+                   ('end', np.float),
+                   ('chord', CHORD_DTYPE)]
+
+NO_CHORD = (-1, -1, np.zeros(12, dtype=np.int))
+UNKNOWN_CHORD = (-1, -1, np.ones(12, dtype=np.int) * -1)
+
+
+def chords(labels):
+    """
+    Transform a list of chord labels into an array of internal numeric
+    representations.
+
+    Parameters
+    ----------
+    labels : list
+        List of chord labels (str)
+
+    Returns
+    -------
+    numpy.array
+        Structured array with columns 'root', 'bass', and 'intervals',
+        containing a numeric representation of chords.
+
+    """
+    crds = np.zeros(len(labels), dtype=CHORD_DTYPE)
+    cache = {}
+    for i, lbl in enumerate(labels):
+        cv = cache.get(lbl, None)
+        if cv is None:
+            cv = chord(lbl)
+            cache[lbl] = cv
+            crds[i] = cv
+        else:
+            crds[i] = cv
+
+    return crds
+
+
+def chord(label):
+    """
+    Transform a chord label into the internal numeric represenation of
+    (root, bass, intervals array).
+
+    Parameters
+    ----------
+    label : str
+        Chord label
+
+    Returns
+    -------
+    tuple
+        Numeric representation of the chord: (root, bass, intervals array)
+
+    """
+    if label == 'N':
+        return NO_CHORD
+    if label == 'X':
+        return UNKNOWN_CHORD
+
+    c_idx = label.find(':')
+    s_idx = label.find('/')
+
+    if c_idx == -1:
+        quality_str = 'maj'
+        if s_idx == -1:
+            root_str = label
+            bass_str = ''
+        else:
+            root_str = label[:s_idx]
+            bass_str = label[s_idx + 1:]
+    else:
+        root_str = label[:c_idx]
+        if s_idx == -1:
+            quality_str = label[c_idx + 1:]
+            bass_str = ''
+        else:
+            quality_str = label[c_idx + 1:s_idx]
+            bass_str = label[s_idx + 1:]
+
+    root = pitch(root_str)
+    bass = interval(bass_str) if bass_str else 0
+    ivs = chord_intervals(quality_str)
+    ivs[bass] = 1
+
+    return root, bass, ivs
+
+
+_l = [0, 1, 1, 0, 1, 1, 1]
+_chroma_id = (np.arange(len(_l) * 2) + 1) + np.array(_l + _l).cumsum() - 1
+
+
+def modify(base, modifier):
+    """
+    Modify a pitch class in integer representation by a given modifier string.
+    A modifier string can be any sequence of 'b' (one semitone down)
+    and '#' (one semitone up).
+
+    Parameters
+    ----------
+    base : int
+        Pitch class as integer
+    modifier : str
+        String of modifiers ('b' or '#')
+
+    Returns
+    -------
+    int
+        Modified root note
+
+    """
+    for m in modifier:
+        if m == 'b':
+            base -= 1
+        elif m == '#':
+            base += 1
+        else:
+            raise ValueError('Unknown modifier: {}'.format(m))
+    return base
+
+
+def pitch(pitch_str):
+    """
+    Converts a string representation of a pitch class (consisting of root
```

Comment:
PEP 257 suggests to use rather "do this" than "does..." style, i.e. "convert...".